airflow/dags/dags.py

```diff
@@ -128,4 +147,8 @@ def sql_enrich_duplicates(schema_tbl, key_columns, order_by_columns):
             "sql_enrich_duplicates": sql_enrich_duplicates,
         },
         user_defined_filters=user_defined_filters,
+        default_args={
+            "on_failure_callback": email_failure,
+            "on_retry_callback": email_failure,
```

Comment:
I kind of worry about emailing on retry since I think some external DAG dependency tasks may "retry" multiple times while waiting for the external dependency to finish.

Reply:
Ah, shoot--good point. Let me double check / tweak to ensure external dependencies won't fire retry messages.

Reply:
> Ah, shoot--good point. Let me double check / tweak to ensure external dependencies won't fire retry messages.

Seems like @machow never got around to checking this, and that's OK. However, let's make sure we're not getting a lot of noise from retries before merging this.

Reply:
Also, do these callbacks override the default method of sending off error emails?

Reply:
@atvaccaro

Reply:
I believe they are separate but I'm not 100% sure; I can test when I have time.

For what it's worth, I have never used email-based alerting with Airflow. I've always used Slack for errors and something like Prometheus to check for positive state.